[2.x] Scala 3.8 REPL support

> [sbt 2.0.0-RC7 merged 1.12.x changes](https://github.com/sbt/sbt/commit/17f8fceed30da74f4a722d5a89eaed00aa8dc7f0) but  does not work
> 
> 
> `build.sbt`
> 
> ```scala
> scalaVersion := "3.8.0-RC2"
> ```
> 
> `project/build.properties`
> 
> ```properties
> sbt.version=2.0.0-RC7
> ```
> 
> 
> ```
> [error] java.lang.NoClassDefFoundError: dotty/tools/repl/ReplDriver
> [error] 	at xsbt.ConsoleInterface.run(ConsoleInterface.java:43)
> [error] 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> [error] 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> [error] 	at sbt.internal.inc.AnalyzingCompiler.invoke(AnalyzingCompiler.scala:329)
> [error] 	at sbt.internal.inc.AnalyzingCompiler.console(AnalyzingCompiler.scala:233)
> [error] 	at sbt.Console.console0$1(Console.scala:73)
> [error] 	at sbt.Console.apply$$anonfun$1$$anonfun$1$$anonfun$1(Console.scala:83)
> [error] 	at sbt.Console.apply$$anonfun$1$$anonfun$1$$anonfun$adapted$1(Console.scala:83)
> [error] 	at scala.Function0.apply$mcV$sp(Function0.scala:42)
> [error] 	at sbt.Run$.executeSuccess(Run.scala:240)
> [error] 	at sbt.Console.apply$$anonfun$1$$anonfun$1(Console.scala:83)
> [error] 	at sbt.internal.util.Terminal.withRawInput(Terminal.scala:159)
> [error] 	at sbt.internal.util.Terminal.withRawInput$(Terminal.scala:23)
> [error] 	at sbt.internal.util.Terminal$ProxyTerminal$.withRawInput(Terminal.scala:454)
> [error] 	at sbt.Console.apply$$anonfun$1(Console.scala:83)
> [error] 	at sbt.internal.util.Terminal$TerminalImpl.withRawOutput(Terminal.scala:1064)
> [error] 	at sbt.internal.util.Terminal$ProxyTerminal$.withRawOutput(Terminal.scala:493)
> [error] 	at sbt.Console.apply(Console.scala:84)
> [error] 	at sbt.Console.apply(Console.scala:51)
> [error] 	at sbt.Console.apply(Console.scala:43)
> [error] 	at sbt.Defaults$.consoleTask$$anonfun$1(Defaults.scala:2054)
> [error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
> [error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
> [error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:79)
> [error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:79)
> [error] 	at sbt.std.Transform$$anon$3.work(Transform.scala:79)
> [error] 	at sbt.std.Transform$$anon$3.work(Transform.scala:79)
> [error] 	at sbt.Execute.submit$$anonfun$1$$anonfun$1(Execute.scala:284)
> [error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
> [error] 	at sbt.Execute.work(Execute.scala:295)
> [error] 	at sbt.Execute.submit$$anonfun$1(Execute.scala:284)
> [error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$2(ConcurrentRestrictions.scala:275)
> [error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:75)
> [error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:73)
> [error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
> [error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
> [error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
> [error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
> [error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
> [error] 	at java.base/java.lang.Thread.run(Thread.java:1583)
> [error] Caused by: java.lang.ClassNotFoundException: dotty.tools.repl.ReplDriver
> [error] 	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
> [error] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:593)
> [error] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
> [error] 	at xsbt.ConsoleInterface.run(ConsoleInterface.java:43)
> [error] 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> [error] 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> [error] 	at sbt.internal.inc.AnalyzingCompiler.invoke(AnalyzingCompiler.scala:329)
> [error] 	at sbt.internal.inc.AnalyzingCompiler.console(AnalyzingCompiler.scala:233)
> [error] 	at sbt.Console.console0$1(Console.scala:73)
> [error] 	at sbt.Console.apply$$anonfun$1$$anonfun$1$$anonfun$1(Console.scala:83)
> [error] 	at sbt.Console.apply$$anonfun$1$$anonfun$1$$anonfun$adapted$1(Console.scala:83)
> [error] 	at scala.Function0.apply$mcV$sp(Function0.scala:42)
> [error] 	at sbt.Run$.executeSuccess(Run.scala:240)
> [error] 	at sbt.Console.apply$$anonfun$1$$anonfun$1(Console.scala:83)
> [error] 	at sbt.internal.util.Terminal.withRawInput(Terminal.scala:159)
> [error] 	at sbt.internal.util.Terminal.withRawInput$(Terminal.scala:23)
> [error] 	at sbt.internal.util.Terminal$ProxyTerminal$.withRawInput(Terminal.scala:454)
> [error] 	at sbt.Console.apply$$anonfun$1(Console.scala:83)
> [error] 	at sbt.internal.util.Terminal$TerminalImpl.withRawOutput(Terminal.scala:1064)
> [error] 	at sbt.internal.util.Terminal$ProxyTerminal$.withRawOutput(Terminal.scala:493)
> [error] 	at sbt.Console.apply(Console.scala:84)
> [error] 	at sbt.Console.apply(Console.scala:51)
> [error] 	at sbt.Console.apply(Console.scala:43)
> [error] 	at sbt.Defaults$.consoleTask$$anonfun$1(Defaults.scala:2054)
> [error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
> [error] 	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
> [error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:79)
> [error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:79)
> [error] 	at sbt.std.Transform$$anon$3.work(Transform.scala:79)
> [error] 	at sbt.std.Transform$$anon$3.work(Transform.scala:79)
> [error] 	at sbt.Execute.submit$$anonfun$1$$anonfun$1(Execute.scala:284)
> [error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
> [error] 	at sbt.Execute.work(Execute.scala:295)
> [error] 	at sbt.Execute.submit$$anonfun$1(Execute.scala:284)
> [error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$2(ConcurrentRestrictions.scala:275)
> [error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:75)
> [error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:73)
> [error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
> [error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
> [error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
> [error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
> [error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
> [error] 	at java.base/java.lang.Thread.run(Thread.java:1583)
> [error] (Compile / console) java.lang.NoClassDefFoundError: dotty/tools/repl/ReplDriver
> ``` 

 _Originally posted by @xuwei-k in [#8348](https://github.com/sbt/sbt/issues/8348#issuecomment-3583474229)_

--- a/main/src/main/scala/sbt/internal/Compiler.scala
+++ b/main/src/main/scala/sbt/internal/Compiler.scala
@@ -165,20 +165,22 @@ object Compiler:
       yield file
 
     val allCompilerJars = toolReport.modules.flatMap(_.artifacts.map(_._2))
-    val allDocJars =
-      fullReport
-        .configuration(Configurations.ScalaDocTool)
-        .map(updateLibraryToCompileConfiguration)
-        .toSeq
-        .flatMap(_.modules)
-        .flatMap(_.artifacts.map(_._2))
+    val extraToolJars = extraToolConf match
+      case Some(extra) =>
+        fullReport
+          .configuration(extra)
+          .map(updateLibraryToCompileConfiguration)
+          .toSeq
+          .flatMap(_.modules)
+          .flatMap(_.artifacts.map(_._2))
+      case None => Nil
     val libraryJars = ScalaArtifacts.libraryIds(sv).flatMap(file)
 
     makeScalaInstance(
       sv,
       libraryJars,
       allCompilerJars,
-      allDocJars,
+      extraToolJars,
       Keys.state.value,
       Keys.scalaInstanceTopLoader.value,
     )

--- a/project/Dependencies.scala
+++ b/project/Dependencies.scala
@@ -12,7 +12,7 @@ object Dependencies {
 
   // sbt modules
   private val ioVersion = nightlyVersion.getOrElse("1.10.5")
-  val zincVersion = nightlyVersion.getOrElse("2.0.0-M9")
+  val zincVersion = nightlyVersion.getOrElse("2.0.0-M10")
 
   private val sbtIO = "org.scala-sbt" %% "io" % ioVersion
 

